### PR TITLE
Perform gradient clipping on global batch when using gradient accumulation

### PR DESCRIPTION
**Pull request by ashors1**
Refactoring to allow gradient clipping to be performed on full batch rather than subbatches when using `ShardedStaticAccumulator`. Note that this refactor allows us to maintain support for `enable_skip_step_on_gradient_anomalies` and requires `x+1` grad norm calculations per global batch when using `ShardedStaticAccumulator` with `x` subbatches (once per subbatch to determine whether step should be skipped, once when applying gradient clipping in base optimizer update) and requires one grad clip per global batch.

This PR should be taken together with the corresponding Paxml PR.

### CODE REVIEW

File: praxis/optimizers.py [hunk 1]
```diff
@@ -2775,8 +2777,54 @@ def update_fn(updates: NestedJTensor,
...
+          return grads, grad_scale
+    
+        raw_grad_norm = _compute_grad_norm(raw_grads)
+        
+        grads, grad_scale = clip_grads(raw_grads, raw_grad_norm)
+        return grads
+    
       averaged_updated = jax.tree_map(lambda acc: acc / num_sub_batches,
                                       new_accumulated_update)
+      averaged_updated = scale_gradients(averaged_updated,
```
Comment (zhangqiaorjc) [inline praxis/optimizers.py:2825]: nit: maybe s/averaged_updated/scaled_updated just to be clear

File: praxis/optimizers.py [hunk 2]
```diff
@@ -2775,8 +2777,54 @@ def update_fn(updates: NestedJTensor,
...
+            assert clip_grad_single_norm_to_value == 0.
+            grad_scale = jnp.minimum(
+                jnp.array(1, grad_norm.dtype),
+                jnp.array(clip_grad_norm_to_value, grad_norm.dtype)
+                / grad_norm)
+            grads = jax.tree_map(lambda g: g * grad_scale, grads)
+          elif clip_grad_single_norm_to_value:
+            assert clip_grad_norm_to_value == 0.
+            grad_single_norm = jax.tree_map(lambda x: jnp.sqrt(jnp.sum(x * x)),
+                                          grads)
```
Comment (zhangqiaorjc) [inline praxis/optimizers.py:2804]: nit: weird indent

File: praxis/optimizers.py [hunk 3]
```diff
@@ -2775,8 +2777,54 @@ def update_fn(updates: NestedJTensor,
...
+            grads = jax.tree_map(scale_gradient, grads, grad_single_norm)
+            grad_scale = jnp.array(1.0)
+          else:
+            # no clipping is needed.
+            grad_scale = jnp.array(1.0)
+          return grads, grad_scale
+    
+        raw_grad_norm = _compute_grad_norm(raw_grads)
+        
+        grads, grad_scale = clip_grads(raw_grads, raw_grad_norm)
```
Comment (zhangqiaorjc) [inline praxis/optimizers.py:2820]: do we need to compute and return `grad_scale`?

File: praxis/optimizers.py [hunk 4]
```diff
@@ -2775,8 +2777,54 @@ def update_fn(updates: NestedJTensor,
...
+                  jnp.array(clip_grad_single_norm_to_value,
+                            norm.dtype) / norm)
+            grads = jax.tree_map(scale_gradient, grads, grad_single_norm)
+            grad_scale = jnp.array(1.0)
+          else:
+            # no clipping is needed.
+            grad_scale = jnp.array(1.0)
+          return grads, grad_scale
+    
+        raw_grad_norm = _compute_grad_norm(raw_grads)
```
Comment (zhangqiaorjc) [inline praxis/optimizers.py:2818]: iiuc, if `clip_grad_single_norm_to_value` is True, then `raw_grad_norm` is not used and we have to compute grad_single_norm separately anyways?

can we move the if-elif-else statement inside out and avoid redundant computation?

**Review by zhangqiaorjc — CHANGES_REQUESTED** [review]
thanks Anna!

File: praxis/optimizers.py (hunk 4 above)
Comment (ashors1) [reply praxis/optimizers.py]: Definitely. I have addressed this with my latest commit

File: praxis/optimizers.py (hunk 3 above)
Comment (ashors1) [reply praxis/optimizers.py]: This is not needed. I no longer return `grad_scale` with the latest commit

File: praxis/optimizers.py [hunk 5]
```diff
@@ -2781,10 +2783,54 @@ def update_fn(updates: NestedJTensor,
...
+      def _compute_grad_norm(grads: NestedMap) -> JTensor:
+        """Computes total grad norm."""
+        grad_norms_squared = jax.tree_map(lambda x: jnp.sum(x * x), grads)
+        grad_norms_squared, _ = jax.tree_util.tree_flatten(grad_norms_squared)
+        return jnp.sqrt(jnp.sum(jnp.stack(grad_norms_squared)))
+
+
+      def scale_gradients(
+          raw_grads: NestedMap,
+          clip_grad_norm_to_value: Optional[float] = None,
```
Comment (zhangqiaorjc) [inline praxis/optimizers.py:2796]: looking at praxis optimizers, clip_gradient_norm_to_value and clip_gradient_single_norm_to_value default are 0.0 and not None right?

so perhaps the types here should be float and default 0.0 instead of Optional?

File: praxis/optimizers.py [hunk 6]
```diff
@@ -2781,10 +2783,54 @@ def update_fn(updates: NestedJTensor,
...
+          clip_grad_single_norm_to_value: Optional[float] = None):
+
+        def clip_grads(grads):
+          if clip_grad_norm_to_value:
```
Comment (zhangqiaorjc) [inline praxis/optimizers.py:2800]: maybe assert only one of them is true?

**Review by zhangqiaorjc — CHANGES_REQUESTED** [review]

**Review by zhangqiaorjc — APPROVED** [review]
